plugins/org.eclipse.n4js.transpiler.dts/src/org/eclipse/n4js/transpiler/dts/transform/ExpandThisFunctionsTransformation.xtend

```diff
@@ -0,0 +1,262 @@
+			if (isSpecConstructor(ctor)) {
+				return ctor;
+			} else if (ctor !== null) {
+				return null;
+			}
+		} while (true);
+	}
+	
+	def N4MethodDeclaration getConstructor(N4ClassDeclaration clazz) {
+		val methods = collectNodes(clazz, false, N4MethodDeclaration);
+		for (method : methods) {
+			if (method.isConstructor) {
+				return method;
+			}
+		}
+		return null;
+	}
+	
+	def FormalParameter getSpecFPar(N4MethodDeclaration method) {
+		if (method === null) {
+			return null;
+		}
+		if (!method.isConstructor) {
+			return null;
+		}
+		if (method.fpars.empty) {
+			return null;
+		}
+		for (fpar : method.fpars) {
+			if (AnnotationDefinition.SPEC.hasAnnotation(fpar)) {
+				return fpar;
+			}
+		}
+		return null;
+	}
+	
+	def boolean isSpecConstructor(TMethod method) {
+		if (method === null) {
+			return false;
+		}
+		if (!method.isConstructor) {
+			return false;
+		}
+		if (method.fpars.empty) {
+			return false;
+		}
+		for (fpar : method.fpars) {
+			if (AnnotationDefinition.SPEC.hasAnnotation(fpar)) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	def N4MethodDeclaration addOverwritingSpecConstructor(TMethod constructor, N4ClassDeclaration clazz) {
+		val clazzOrig = state.tracer.getOriginalASTNode(clazz) as N4ClassDeclaration;
+		val newConstructor = _N4MethodDecl("constructor", new ArrayList());
+		clazz.ownedMembersRaw += newConstructor;
+		for (fpar: constructor.fpars) {
+			val newFPar = _FormalParameter(fpar.name);
+			newConstructor.fpars += newFPar;
+			var typeRef = fpar.typeRef;
+			if (AnnotationDefinition.SPEC.hasAnnotation(fpar)) {
+				typeRef = getConcreteTypeRef(fpar.typeRef, clazzOrig.definedType);
+			}
+			
+			newFPar.declaredTypeRefNode = _TypeReferenceNode(state, typeRef);
+			state.info.setOriginalProcessedTypeRef(newFPar.declaredTypeRefNode, typeRef);
+		}
+		return newConstructor;
+	}
+
+	def N4MethodDeclaration addOverwritingStaticMethod(TMethod method, N4ClassDeclaration clazz) {
+		val clazzOrig = state.tracer.getOriginalASTNode(clazz) as N4ClassDeclaration;
+		val newMethod = _N4MethodDecl(method.name);
+		newMethod.declaredModifiers += N4Modifier.STATIC;
+		clazz.ownedMembersRaw += newMethod;
+		for (fpar: method.fpars) {
+			val newFPar = _FormalParameter(fpar.name);
+			newMethod.fpars += newFPar;
+			var typeRef = fpar.typeRef;
+			newFPar.declaredTypeRefNode = _TypeReferenceNode(state, typeRef);
+			state.info.setOriginalProcessedTypeRef(newFPar.declaredTypeRefNode, typeRef);
+		}
+		
+
+		Preconditions.checkState(method.returnTypeRef instanceof ThisTypeRef);
+		
+		val typeRef = getConcreteTypeRef(method.returnTypeRef, clazzOrig.definedType);
+		newMethod.declaredReturnTypeRefNode = _TypeReferenceNode(state, typeRef);
+		state.info.setOriginalProcessedTypeRef(newMethod.declaredReturnTypeRefNode, typeRef);
+		
+		return newMethod;
+	}
+	
+	def private TypeRef getConcreteTypeRef(TypeRef typeRef, Type targetType) {
+		val thisTypeRef = TypeUtils.createTypeRefWithParamsAsArgs(targetType);
```

Comment:
Since the type created here is used for static methods, I wonder if wildcards have to be used here instead of the "params" (i.e. the type parameters of 'targetType'), which are unbound in the static context.

(method `TypeUtils#createTypeRef(Type declaredType, TypingStrategy typingStrategy, boolean autoCreateTypeArgs, TypeArgument... typeArgs)` could be used for that; it is creating wildcards if no type arguments are provided)